Multi transport logging

MattMorgis opened the pull request:
Allows for a config like this:
```yaml
logger:
  transports: [file, console]
  level: debug
  path_pattern: "logs/mcp-agent-{timestamp}.jsonl"
  timestamp_format: "%Y%m%d_%H%M%S"
```

Keeps backward compatibility: 
```yaml
logger:
  type: file
  level: debug
  path: "logs/mcp-agent.jsonl"
```

saqadri commented on src/mcp_agent/logging/transport.py:
```diff
@@ -432,19 +434,113 @@ async def _process_events(self):
     settings: LoggerSettings, event_filter: EventFilter | None = None
 ) -> EventTransport:
     """Create event transport based on settings."""
+    # Check if multiple transports are configured
+    if hasattr(settings, "transports") and len(settings.transports) >= 1:
+        transports: List[EventTransport] = []
+
+        for transport_type in settings.transports:
+            if transport_type == "none":
+                continue
+            elif transport_type == "console":
+                transports.append(ConsoleTransport(event_filter=event_filter))
+            elif transport_type == "file":
+                filepath = generate_log_filename(settings)
+                transports.append(
+                    FileTransport(filepath=filepath, event_filter=event_filter)
+                )
+            elif transport_type == "http":
+                if not settings.http_endpoint:
+                    print(
+                        "Warning: HTTP endpoint required for HTTP transport. Skipping."
+                    )
+                    continue
+                transports.append(
+                    HTTPTransport(
+                        endpoint=settings.http_endpoint,
+                        headers=settings.http_headers,
+                        batch_size=settings.batch_size,
+                        timeout=settings.http_timeout,
+                        event_filter=event_filter,
+                    )
+                )
+        if transports:
+            return MultiTransport(transports)
+        else:
+            return NoOpTransport(event_filter=event_filter)
+    # Backward compatibility for single transport configuration
```
A couple of issues I noticed:

1. if someone specifies just one item in `transports` then we should use that single transport the same way if someone specified `transport`.
2. 'path_pattern' only seems to apply for multiple transports, but if someone specifies just transport type it won't ever apply.
3. Looks like some raise ValueError were changed to warnings. I actually think it's better to raise if some transports are misconfigured.

Also, there's a bit of code duplication happening so we can simplify the entire function as follows:

    transports: List[EventTransport] = []

    transport_types = []

    if hasattr(settings, "transports"):
        transport_types = settings.transports
    else:
        transport_types = [settings.type]
    
    for transport_type in transport_types:
        if transport_type == "none":
                continue
        elif transport_type == "console":
            transports.append(ConsoleTransport(event_filter=event_filter))
        elif transport_type == "file":
            filepath = get_log_filename(settings) # IMPORTANT: needs to include
            #if settings.path:
            #   return settings.path
            #elif settings.path_settings:
            #    <current logic for generate_log_filename>
            #else:
            #    raise ValueError("File path required for file transport. Either specify 'path' or 'path_settings'")
            
            transports.append(
                FileTransport(filepath=filepath, event_filter=event_filter)
            )
        elif transport_type == "http":
            if not settings.http_endpoint:
                raise ValueError("HTTP endpoint required if HTTP transport is specified")
            transports.append(
                HTTPTransport(
                    endpoint=settings.http_endpoint,
                    headers=settings.http_headers,
                    batch_size=settings.batch_size,
                    timeout=settings.http_timeout,
                    event_filter=event_filter,
                )
            )
        else:
            raise ValueError(f"Unsupported transport type: {settings.type}")#
        
    if not transports:
        return NoOpTransport(event_filter=event_filter)
    elif len(transports) > 1:
        return MultiTransport(transports)
    else:
        return transports[0]

saqadri reviewed (changes requested):
Thanks for adding this @MattMorgis! Directionally looks good to me, but I have a few suggestions which would be great to incorporate before merging. Please let me know if you have any questions on them, and thanks again!

saqadri commented on src/mcp_agent/logging/transport.py:
```diff
@@ -432,19 +434,113 @@ async def _process_events(self):
+                print(f"  {transport.__class__.__name__}: {exc}")
+
+
+def generate_log_filename(settings: LoggerSettings) -> str:
```
Some comments in other parts of the change would affect this function, so just flagging for completeness:

This function should take into account if `path` is specified, in which case it should just return that

```
#if settings.path:
        #   return settings.path
        #elif settings.path_settings:
        #    <current logic for generate_log_filename>
        #else:
        #    raise ValueError("File path required for file transport. Either specify 'path' or 'path_settings'")
 
```

MattMorgis commented on src/mcp_agent/logging/transport.py:
```diff
@@ -432,19 +434,113 @@ async def _process_events(self):
+
+    # Generate timestamp or run ID
+    if hasattr(settings, "use_run_id") and settings.use_run_id:
+        replacement = str(uuid.uuid4())
```
#47

MattMorgis commented:
@saqadri I think I addressed all of your feedback. I also updated the README and the basic agent example to reflect the new config options. Let me know if you need anything else!

MattMorgis commented on src/mcp_agent/config.py:
```diff
@@ -183,6 +213,12 @@ class LoggerSettings(BaseModel):
     """Path to log file, if logger 'type' is 'file'."""
 
+    # Settings for advanced log path configuration
+    path_settings: Optional["LogPathSettings"] = None
```
Ha, nice catch. A combination of Claude doing it + me still new enough at Python to not notice. I can submit a follow up, the rest of the file uses the `| None` pattern
